# Quick test

In [109]:
# required Python imports
import numpy as np

In [110]:
# define a function that takes an estimated T value and a range of theta and phi values and return the difference
# it assumes that Ta, Tb, and Tc values are those of forsterite (90, 50, 20)
def func(T, theta, phi):
    """ TODO
    """
    left_term = 90 * np.cos(theta)**2 * np.sin(phi)**2 + \
                50 * np.sin(theta)**2 * np.sin(phi)**2 + \
                20 * np.cos(phi)**2
       
    return (left_term - T)**2

In [111]:
# generate a mesh of spherical/polar coordinates
theta = np.linspace(0, 2*np.pi, 200)
phi = np.arccos(1 - 2 * np.linspace(0, 1, 200))  # in this case np.linspace goes from 0 to 1 (not pi)!
phi, theta = np.meshgrid(phi, theta)

# set the "experimentally measured" T value
T = 45.97

# calculate the difference
results = func(T, theta, phi)

In [112]:
print(results.min(), results.max()) 

2.364320108358796e-05 1938.485245439112


In [128]:
# check for results that are close to 0 creating a mask
mask = np.isclose(0, results, rtol=1e-05, atol=5e-05)
np.any(mask)

True

In [129]:
# use the mask to find which phi values met the condition
np.rad2deg(phi[mask])

array([ 68.47949562, 111.52050438,  68.47949562, 111.52050438])

In [130]:
np.rad2deg(theta[mask])

array([ 90.45226131,  90.45226131, 269.54773869, 269.54773869])

So we obtain, as expected, four different solutions for a single data (in degrees): $(68.5, 90.5)$, $(111.5, 90.5)$, $(68.5, 269.5)$, and $(111.5, 269.5)$

In [131]:
# just check
func(40.32, np.deg2rad(90.45), np.deg2rad(55.40))

6.849545254751173e-05

In [135]:
# set a new "experimentally measured" T value
T2 = 70.16

# calculate the difference
results2 = func(T2, theta, phi)

# check for results that are close to 0 creating a mask
mask2 = np.isclose(0, results2, rtol=1e-05, atol=5e-06)

# use the mask to find which phi and theta values met the condition
np.rad2deg(phi[mask2])

array([ 58.15398464, 121.84601536,  58.15398464, 121.84601536])

In [136]:
np.rad2deg(theta[mask2])

array([173.66834171, 173.66834171, 186.33165829, 186.33165829])

In [146]:
# set a new "experimentally measured" T value
T3 = 22.80

# calculate the difference
results3 = func(T3, theta, phi)

# check for results that are close to 0 creating a mask
mask3 = np.isclose(0, results3, rtol=1e-05, atol=1.2e-04)

# use the mask to find which phi and theta values met the condition
np.rad2deg(phi[mask3])

array([ 14.10535405, 165.89464595,  14.10535405, 165.89464595])

In [147]:
np.rad2deg(theta[mask3])

array([ 48.84422111,  48.84422111, 311.15577889, 311.15577889])